File: win-prog/ProgWinVB/Pages and Transforms/WhatSizeTransform/WhatSizeTransform.vb
'--------------------------------------------------
' WhatSizeTransform.vb (c) 2002 by Charles Petzold
'--------------------------------------------------
Imports System
Imports System.Drawing
Imports System.Drawing.Drawing2D
Imports System.Windows.Forms

Class WhatSizeTransform
    Inherits PrintableForm

    Shared Shadows Sub Main()
        Application.Run(new WhatSizeTransform())
    End Sub

    Sub New()
        Text = "What Size? With TransformPoints"
    End Sub

    Protected Overrides Sub DoPage(ByVal grfx As Graphics, ByVal clr As Color, ByVal cx As Integer, ByVal cy As Integer)
        Dim br As New SolidBrush(clr)
        Dim y As Integer = 0
        Dim apt() As Point = {New Point(cx, cy)}

        grfx.TransformPoints(CoordinateSpace.Device, _
                         CoordinateSpace.Page, apt)

        DoIt(grfx, br, y, apt(0), GraphicsUnit.Pixel)
        DoIt(grfx, br, y, apt(0), GraphicsUnit.Display)
        DoIt(grfx, br, y, apt(0), GraphicsUnit.Document)
        DoIt(grfx, br, y, apt(0), GraphicsUnit.Inch)
        DoIt(grfx, br, y, apt(0), GraphicsUnit.Millimeter)
        DoIt(grfx, br, y, apt(0), GraphicsUnit.Point)
    End Sub

    Private Sub DoIt(ByVal grfx As Graphics, ByVal br As Brush, ByRef y As Integer, ByVal pt As Point, ByVal gu As GraphicsUnit)
        Dim gs As GraphicsState = grfx.Save()

        grfx.PageUnit  = gu
        grfx.PageScale = 1

        Dim aptf() As PointF = {Point.op_Implicit(pt)}
        grfx.TransformPoints(CoordinateSpace.Page, _
                         CoordinateSpace.Device, aptf)

        Dim szf As New SizeF(aptf(0))
        grfx.Restore(gs)
        grfx.DrawString(gu.ToString() & ": " & szf.ToString(), Font, br, 0, y)
        y += CInt(Math.Ceiling(Font.GetHeight(grfx)))
    End Sub
End Class

File: win-prog/ProgWinCS/Pages and Transforms/WhatSizeTransform/WhatSizeTransform.cs
//------------------------------------------------
// WhatSizeTransform.cs © 2001 by Charles Petzold
//------------------------------------------------
using System;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Windows.Forms;

class WhatSizeTransform: PrintableForm
{
     public new static void Main()
     {
          Application.Run(new WhatSizeTransform());
     }
     public WhatSizeTransform()
     {
          Text = "What Size? With TransformPoints";
     }
     protected override void DoPage(Graphics grfx, Color clr, int cx, int cy)
     {
          Brush   brush = new SolidBrush(clr);
          int     y     = 0;
          Point[] apt   = { new Point(cx, cy) };

          grfx.TransformPoints(CoordinateSpace.Device, 
                               CoordinateSpace.Page, apt);

          DoIt(grfx, brush, ref y, apt[0], GraphicsUnit.Pixel);
          DoIt(grfx, brush, ref y, apt[0], GraphicsUnit.Display);
          DoIt(grfx, brush, ref y, apt[0], GraphicsUnit.Document);
          DoIt(grfx, brush, ref y, apt[0], GraphicsUnit.Inch);
          DoIt(grfx, brush, ref y, apt[0], GraphicsUnit.Millimeter);
          DoIt(grfx, brush, ref y, apt[0], GraphicsUnit.Point);
     }
     void DoIt(Graphics grfx, Brush brush, ref int y, 
               Point pt, GraphicsUnit gu)
     {
          GraphicsState gs = grfx.Save();

          grfx.PageUnit  = gu;
          grfx.PageScale = 1;

          PointF[] aptf = { pt };

          grfx.TransformPoints(CoordinateSpace.Page, 
                               CoordinateSpace.Device, aptf);

          SizeF sizef = new SizeF(aptf[0]);
          grfx.Restore(gs);

          grfx.DrawString(gu + ": " + sizef, Font, brush, 0, y);
          y += (int) Math.Ceiling(Font.GetHeight(grfx));
     }
}
